open-apparel-registry/open-apparel-registry
# Move the facility match confirm/reject API to allow single-item…

## Overview

To support confirming and rejecting matches for single-item submissions, we need to move the confirm/reject API out from under the facility list API. 

To keep the number of UI changes to a minimum we are returning the same list item details when a request is successful.

We attempt to make it easy for API consumers to complete a partial match by constructing the full paths that can be POSTed to to confirm or reject the match.

Connects #916

## Testing Instructions

* Run `./scripts/resetdb`
* Run `./scripts/manage user_groups -e c8@example.com -a add -g can_submit_facility`
* Log in as `c8@example.com:password`.
* Browse http://localhost:6543/profile/8. Generate an API token and export it into the environment
```
export OAR_API_TOKEN={token}
```
* Submit a facility. Verify that the request is successful and that a `POTENTIAL_MATCH` is returned with with `confirm_match_url` and `reject_match_url` keys. 
```
http POST http://localhost:8081/api/facilities/ \
"Authorization: Token $OAR_API_TOKEN" \
<<< '{
  "country": "China",
  "name": "Huai An Yuan Tong Headwear Mfg. Co. Ltd.",
  "address": "No.1 Yan Huang Avenue Lian Shui New Industrial Zone,Huai An,Jiangsu 223400. China"
}'
```
* Take the OAR from the response and browse http://localhost:6543/facilities/{oar_id}. Verify that the currently logged in contributor is NOT yet listed as a facility contributor.
* Copy the `confirm_match_url` and POST to it. Verify a successful response with a `CONFIRMED_MATCH` status and verify that the currently logged in contributor is now listed on the facility details page http://localhost:6543/facilities/{oar_id}
```
http POST http://localhost:8081/api/facility-matches/934/confirm/ \
"Authorization: Token $OAR_API_TOKEN"
```
* Submit the same details again. Confirm that another `POTENTIAL_MATCH` is created.
```
http POST http://localhost:8081/api/facilities/ \
"Authorization: Token $OAR_API_TOKEN" \
<<< '{
  "country": "China",
  "name": "Huai An Yuan Tong Headwear Mfg. Co. Ltd.",
  "address": "No.1 Yan Huang Avenue Lian Shui New Industrial Zone,Huai An,Jiangsu 223400. China"
}'
```
* POST to the rejection URL. Verify that the successful response contains 2 match records, one `REJECTED` and one `CONFIRMED` with a confidence of `1.0` representing the newly created facility. 
```
http POST http://localhost:8081/api/facility-matches/935/reject/ \
"Authorization: Token $OAR_API_TOKEN"
```
* Browse http://localhost:6543/lists/8
  * Reject the match for row index 10. Verify that a new facility is created.
  * Confirm the match for row index 11. Verify that the currently logged in contributor is listed on the details page for the facility.

## Checklist
